File: tools/_dispatch.py
from typing import Callable
from fitness_core.errors import DomainError
from tools._models import MethodResult, SkippedMethod

MethodOutput = tuple[float, dict | None]


def run_methods(
    requested: list[str],
    explicit: bool,
    compute_fn: Callable[[str], "MethodOutput | None"],
    unit: str,
    *,
    reason_fn: Callable[[str], str] | None = None,
    ndigits: int = 1,
) -> tuple[list[MethodResult], list[SkippedMethod]]:
    """Run each requested method through compute_fn, building per-method results.

    compute_fn(method) returns (value, detail) to include the method, or None to
    indicate its required inputs are missing. Under explicit=True a missing method
    raises DomainError; under explicit=False it is recorded in `skipped`. compute_fn
    may itself raise DomainError for an unknown method. Duplicate methods are ignored.
    """
    results: list[MethodResult] = []
    skipped: list[SkippedMethod] = []
    seen: set[str] = set()
    for m in requested:
        if m in seen:
            continue
        seen.add(m)
        out = compute_fn(m)
        if out is None:
            reason = reason_fn(m) if reason_fn else f"{m}: required inputs missing"
            if explicit:
                raise DomainError(reason)
            skipped.append(SkippedMethod(method=m, reason=reason))
            continue
        value, detail = out
        results.append(MethodResult(method=m, value=round(value, ndigits),
                                    unit=unit, detail=detail))
    return results, skipped

File: packages/core/src/dispatch.ts
import { DomainError } from "./errors.js";
import type { MethodResult, SkippedMethod } from "./models.js";
import { roundTo } from "./math/stats.js";

export type MethodOutput = [number, Record<string, unknown> | null] | null;
export type ComputeFn = (method: string) => MethodOutput;

export interface RunMethodsOptions {
  reasonFn?: (method: string) => string;
  ndigits?: number;
}

export function runMethods(
  requested: string[],
  explicit: boolean,
  computeFn: ComputeFn,
  unit: string,
  opts: RunMethodsOptions = {},
): { results: MethodResult[]; skipped: SkippedMethod[] } {
  const { reasonFn, ndigits = 1 } = opts;
  const results: MethodResult[] = [];
  const skipped: SkippedMethod[] = [];
  const seen = new Set<string>();
  for (const m of requested) {
    if (seen.has(m)) continue;
    seen.add(m);
    const out = computeFn(m);
    if (out === null) {
      const reason = reasonFn ? reasonFn(m) : `${m}: required inputs missing`;
      if (explicit) throw new DomainError(reason);
      skipped.push({ method: m, reason });
      continue;
    }
    const [value, detail] = out;
    results.push({ method: m, value: roundTo(value, ndigits), unit, detail });
  }
  return { results, skipped };
}
